block/block.go

```diff
@@ -1,13 +1,42 @@
 package block
 
 import (
+	"context"
+	"errors"
 	"fmt"
 
+	"github.com/dymensionxyz/dymint/fraud"
+
 	errorsmod "cosmossdk.io/errors"
 
 	"github.com/dymensionxyz/dymint/types"
 )
 
+// applyBlockWithFraudHandling calls applyBlock and validateBlockBeforeApply with fraud handling.
+func (m *Manager) applyBlockWithFraudHandling(block *types.Block, commit *types.Commit, blockMetaData types.BlockMetaData, validate bool) error {
+	if validate {
+		if err := m.validateBlockBeforeApply(block, commit); err != nil {
+			if errors.Is(err, fraud.ErrFraud) {
+				m.FraudHandler.HandleFault(context.Background(), err)
+			} else if err != nil {
+				m.blockCache.Delete(block.Header.Height)
+				// TODO: can we take an action here such as dropping the peer / reducing their reputation?
+			}
+
+			return fmt.Errorf("block not valid at height %d, dropping it: err:%w", block.Header.Height, err)
+		}
+	}
+
+	if err := m.applyBlock(block, commit, blockMetaData); err != nil {
+		if errors.Is(err, fraud.ErrFraud) {
```

Comment:
but applyBlock can also return ErrFraud? is it not only in validateBlockBeforeApply?

Reply:
but isn't consensus messages fraud the same concept as app hash fraud?